File: src/main/java/at/searles/parsing/Parser.java
package at.searles.parsing;

import at.searles.lexer.Tokenizer;
import at.searles.parsing.annotation.AnnotationParser;
import at.searles.parsing.combinators.ParserOrParser;
import at.searles.parsing.combinators.ParserThenRecognizer;
import at.searles.parsing.combinators.ParserThenReducer;
import at.searles.parsing.combinators.ParserToReducer;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import at.searles.parsing.tokens.TokenParser;
import at.searles.regex.Regex;
import org.jetbrains.annotations.Nullable;

/**
 * A parser is an object that reads tokens from a parser and
 * returns an object. Additionally it uses a env object
 * for feedback on errors.
 * <p>
 * Parsers are invertible. A syntax tree (a tree that is easily converted
 * into a string) is returned by the inverse. For a proper inversion,
 * a set is maintained to avoid cycles.
 *
 * @param <T>
 */
public interface Parser<T> extends Recognizable {

    /**
     * Token: lexer::regex^? ~ mapping. In fact, the value is always
     * an instance of FrameStream.Frame (this knowledge can be used
     * to obtain the position of the token).
     */
    static <T> Parser<T> fromToken(int tokenId, Tokenizer tokenizer, boolean exclusive, Mapping<CharSequence, T> mapping) {
        return new TokenParser<>(tokenId, tokenizer, exclusive, mapping);
    }

    /**
     * Token: lexer::regex^? ~ mapping. In fact, the value is always
     * an instance of FrameStream.Frame (this knowledge can be used
     * to obtain the position of the token).
     */
    static <T> Parser<T> fromRegex(Regex regex, Tokenizer tokenizer, boolean exclusive, Mapping<CharSequence, T> mapping) {
        int tokenId = tokenizer.add(regex);
        return fromToken(tokenId, tokenizer, exclusive, mapping);
    }

    /**
     * Returns the parsed value.
     *
     * @param stream The parser stream from which items are read.
     * @return An instance of T or null if this parser cannot be used.
     */
    @Nullable
    T parse(ParserStream stream);

    default @Nullable
    ConcreteSyntaxTree print(T t) {
        throw new UnsupportedOperationException("printing not supported");
    }

    /**
     * A B
     */
    default <U> Parser<U> then(Reducer<T, U> reducer) {
        return then(reducer, false);
    }

    default <U> Parser<U> then(Reducer<T, U> reducer, boolean allowParserBacktrack) {
        return then(reducer, allowParserBacktrack, false);
    }

    default <U> Parser<U> then(Reducer<T, U> reducer, boolean allowParserBacktrack, boolean allowPrinterBacktrack) {
        return new ParserThenReducer<>(this, reducer, allowParserBacktrack, allowPrinterBacktrack);
    }

    default Parser<T> then(Recognizer recognizer) {
        return then(recognizer, false);
    }

    default Parser<T> then(Recognizer recognizer, boolean allowParserBacktrack) {
        return new ParserThenRecognizer<>(this, recognizer, allowParserBacktrack);
    }

    /**
     * A | B
     */
    default Parser<T> or(Parser<T> alternative) {
        return or(alternative, false);
    }

    /**
     * scala: A.|(B, true)
     */
    default Parser<T> or(Parser<T> alternative, boolean swapOnInvert) {
        return new ParserOrParser<>(this, alternative, swapOnInvert);
    }

    /**
     * A > fold
     */
    default <L, V> Reducer<L, V> fold(Fold<L, T, V> fold) {
        return new ParserToReducer<>(this, fold);
    }

    /**
     * Useful to simplify parser's toString output. Automatically
     * created for each rule.
     */
    default Parser<T> ref(String label) {
        return new Ref<T>(label).set(this);
    }

    /**
     * A @ annotation
     * \
     */
    default <M> Parser<T> annotate(M annotation) {
        return new AnnotationParser<>(annotation, this);
    }
}

File: src/main/java/at/searles/parsing/Parser.kt
package at.searles.parsing

import at.searles.lexer.Tokenizer
import at.searles.parsing.Reducer.Companion.rep
import at.searles.parsing.combinators.*
import at.searles.parsing.combinators.ext.ParserOrParserReversePrintOrder
import at.searles.parsing.printing.ConcreteSyntaxTree
import at.searles.parsing.ref.RefParser
import at.searles.parsing.tokens.TokenParser
import at.searles.parsingtools.common.PairCreator
import at.searles.parsingtools.list.EmptyListCreator
import at.searles.parsingtools.list.ListAppender
import at.searles.parsingtools.list.ListCreator
import at.searles.parsingtools.opt.NoneCreator
import at.searles.parsingtools.opt.SomeCreator
import at.searles.regexp.Regexp
import java.util.*

/**
 * A parser is an object that reads tokens from a parser and
 * returns an object. Additionally it uses a env object
 * for feedback on errors.
 *
 *
 * Parsers are invertible. A syntax tree (a tree that is easily converted
 * into a string) is returned by the inverse. For a proper inversion,
 * a set is maintained to avoid cycles.
 *
 * @param <T>
</T> */
interface Parser<T>: CanRecognize {
    /**
     * Returns the parsed value.
     *
     * Contract: If the parser is successful, stream's interval is set to the position of the
     * successful parse.
     *
     * @param stream The parser stream from which items are read.
     * @return An instance of T or null if this parser cannot be used.
     */
    fun parse(stream: ParserStream): T?

    override fun recognize(stream: ParserStream): Boolean

    fun print(item: T): ConcreteSyntaxTree? {
        throw UnsupportedOperationException("printing not supported")
    }

    fun rep1(separator: Recognizer): Parser<List<T>> {
        return (this + ListCreator()) + (separator + this.plus(ListAppender(1))).rep()
    }

    fun rep1(): Parser<List<T>> {
        return (this + ListCreator()) + this.plus(ListAppender(1)).rep()
    }

    fun rep(separator: Recognizer): Parser<List<T>> {
        return EmptyListCreator<T>() + separator.join(this.plus(ListAppender(0)))
    }

    fun rep(): Parser<List<T>> {
        return EmptyListCreator<T>() + this.plus(ListAppender(0)).rep()
    }

    fun opt(): Parser<Optional<T>> {
        return this + SomeCreator() or NoneCreator()
    }

    infix fun or(other: Parser<T>): Parser<T> {
        return ParserOrParser(this, other)
    }

    infix fun orSwapOnPrint(other: Parser<T>): Parser<T> {
        return ParserOrParserReversePrintOrder(this, other)
    }

    operator fun <L, V> plus(fold: Fold<L, T, V>): Reducer<L, V> {
        return ParserToReducer(this, fold)
    }

    operator fun <U> plus(right: Reducer<T, U>): Parser<U> {
        return ParserThenReducer(this, right)
    }

    operator fun plus(right: Recognizer): Parser<T> {
        return this + right.toReducer()
    }

    operator fun <U> plus(right: Parser<U>): Parser<Pair<T, U>> {
        return this + right.plus(PairCreator())
    }

    /**
     * Useful to simplify parser's toString output. Automatically
     * created for each rule.
     */
    fun ref(label: String): Parser<T> {
        return RefParser<T>(label).apply {
            ref = this@Parser
        }
    }

    fun <L> toPair(): Reducer<L, Pair<L, T>> {
        return this + PairCreator()
    }

    companion object {
        /**
         * Token: lexer::regex^? ~ mapping. In fact, the value is always
         * an instance of FrameStream.Frame (this knowledge can be used
         * to obtain the position of the token).
         */
        fun <T> fromToken(tokenId: Int, tokenizer: Tokenizer, mapping: Mapping<CharSequence, T>): Parser<T> {
            return TokenParser(tokenId, tokenizer) + mapping
        }

        /**
         * Token: lexer::regex^? ~ mapping. In fact, the value is always
         * an instance of FrameStream.Frame (this knowledge can be used
         * to obtain the position of the token).
         */
        fun <T> fromRegex(regexp: Regexp, tokenizer: Tokenizer, mapping: Mapping<CharSequence, T>): Parser<T> {
            val tokenId = tokenizer.add(regexp)
            return fromToken(tokenId, tokenizer, mapping)
        }

        fun <T> Parser<List<T>>.orEmpty(): Parser<List<T>> {
            return this orSwapOnPrint EmptyListCreator()
        }
    }
}